grabbit/src/main/groovy/com/twcable/grabbit/jcr/ProtoPropertyDecorator.groovy

```diff
@@ -0,0 +1,92 @@
+package com.twcable.grabbit.jcr
+
+import com.twcable.grabbit.DateUtil
+import com.twcable.grabbit.proto.NodeProtos.Property as ProtoProperty
+import com.twcable.grabbit.proto.NodeProtos.Value as ProtoValue
+import groovy.transform.CompileStatic
+import org.apache.jackrabbit.value.ValueFactoryImpl
+
+import javax.annotation.Nonnull
+import javax.jcr.Node as JCRNode
+import javax.jcr.Property
+import javax.jcr.PropertyType
+import javax.jcr.Value
+import javax.jcr.ValueFormatException
+
+import static org.apache.jackrabbit.JcrConstants.JCR_MIXINTYPES
+import static org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE
+
+@CompileStatic
+class ProtoPropertyDecorator {
+
+    @Delegate
+    ProtoProperty innerProtoProperty
+
+    ProtoPropertyDecorator(@Nonnull ProtoProperty protoProperty) {
+        this.innerProtoProperty = protoProperty
+    }
+
+
+    void writeToNode(@Nonnull JCRNode node) {
+        if(primaryType || mixinType) throw new IllegalStateException("Refuse to write jcr:primaryType or jcr:mixinType as normal properties.  These are not allowed")
+        try {
+            if (!innerProtoProperty.hasValues()) {
+                node.setProperty(innerProtoProperty.name, getPropertyValue(), innerProtoProperty.type)
+            }
+            else {
+                Value[] values = getPropertyValues()
+                node.setProperty(innerProtoProperty.name, values, innerProtoProperty.type)
+            }
+        }
+        catch (ValueFormatException ex) {
+            if (ex.message.contains("Multivalued property can not be set to a single value")) {
+                //If this is the exception, that means that a property with the name already exists
+                final Property currentProperty = node.getProperty(innerProtoProperty.name)
+                if (currentProperty.multiple) {
+                    final Value[] values = [getPropertyValue()]
+                    node.setProperty(innerProtoProperty.name, values, innerProtoProperty.type)
+                }
+            }
+            else if (ex.message.contains("Single-valued property can not be set to an array of values")) {
+                node.setProperty(innerProtoProperty.name, getPropertyValues().first(), innerProtoProperty.type)
```

Comment:
```groovy
}
catch(ValueFormatException) {
    //Someone's being a pain
    final opinionatedProperty = node.getProperty(innerProtoProperty.name)
    //Are we being rebellious by trying to write a different type?  We should persist our changes?
    if(opinionatedProperty.type != innerProtoProperty.type) {
        //Away with it
        opinionatedProperty.remove()
        return this.writeToNode(node)    
    }
    //If we are trying to write to a property that accepts multiple values, lets succeed to its wishes
    if(opinionatedProperty.multiple) {

    }
    //Or act on the inverse
    else {

    }
}
```
might do it?  According to the documentation, these two cases are the only cases :

"if <code>value</code> cannot be converted to the specified type or if the property already exists and is multi-valued."